File: c/rtmux.cc
#include <fstream>
#include <string>
#include <vector>
#include <optional>
#include <iostream>

#include <nlohmann/json.hpp>

using json = nlohmann::json;

struct Ql {
    std::string ts;
    std::string lvl;
    std::string msg;
    std::optional<json> ctx;
};

enum class RtErr {
    Io,
    Parse
};

bool rx(const Ql& q) {
    if (q.lvl == "ERROR") {
        return true;
    }

    if (q.msg.find("timeout") != std::string::npos ||
        q.msg.find("mismatch") != std::string::npos) {
        return true;
    }

    if (q.ctx.has_value()) {
        const auto& c = q.ctx.value();
        if (c.contains("flag") && c["flag"].is_boolean() && c["flag"] == true) {
            return true;
        }
    }

    return false;
}

std::optional<Ql> px(const std::string& ln) {
    try {
        auto j = json::parse(ln);

        Ql q;
        q.ts  = j.value("timestamp", "");
        q.lvl = j.value("level", "");
        q.msg = j.value("message", "");

        if (j.contains("context")) {
            q.ctx = j["context"];
        }

        return q;
    } catch (...) {
        return std::nullopt;
    }
}

std::pair<size_t, RtErr> rm(const std::string& path) {
    std::ifstream in(path);
    if (!in.is_open()) {
        return {0, RtErr::Io};
    }

    std::string ln;
    size_t n = 0;

    while (std::getline(in, ln)) {
        if (ln.empty()) {
            continue;
        }

        auto q_opt = px(ln);
        if (!q_opt.has_value()) {
            continue;
        }

        const Ql& q = q_opt.value();

        if (rx(q)) {
            // --- placeholder routing cases ---

            // case A: escalate outward
            // if (q.ctx && q.ctx->contains("priority") && (*q.ctx)["priority"] == 10) {
            //     push_ext(q);
            //     continue;
            // }

            // case B: snapshot capture
            // if (q.lvl == "WARN" && q.msg.find("drift") != std::string::npos) {
            //     persist_snapshot(q);
            //     continue;
            // }

            // case C: workflow trigger
            // if (q.ctx && q.ctx->contains("tag") && (*q.ctx)["tag"] == "reconcile") {
            //     trigger_wf(q);
            //     continue;
            // }

            // case D: aggregation bucket
            // if (q.msg.find("settlement") != std::string::npos) {
            //     agg_push(q);
            //     continue;
            // }

            // fallback: no-op
            ++n;
        }
    }

    return {n, RtErr{}};
}

// --- stubs ---
//
// void push_ext(const Ql& q) { /* ... */ }
// void persist_snapshot(const Ql& q) { /* ... */ }
// void trigger_wf(const Ql& q) { /* ... */ }
// void agg_push(const Ql& q) { /* ... */ }

File: rust/rt_mux.rs
use chrono::{DateTime, Utc};
use serde::Deserialize;
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::path::Path;

#[derive(Debug)]
pub enum RtErr {
    Io,
    Parse,
}

#[derive(Debug, Deserialize)]
struct Ql {
    timestamp: DateTime<Utc>,
    level: String,
    message: String,
    context: Option<serde_json::Value>,
}

pub fn rm(p: &Path) -> Result<usize, RtErr> {
    let f = File::open(p).map_err(|_| RtErr::Io)?;
    let r = BufReader::new(f);

    let mut n = 0usize;

    for ln in r.lines() {
        let ln = ln.map_err(|_| RtErr::Io)?;
        if ln.trim().is_empty() {
            continue;
        }

        let q: Ql = match serde_json::from_str(&ln) {
            Ok(v) => v,
            Err(_) => continue,
        };

        if rx(&q) {
            // --- placeholder routing cases ---

            // case A: escalate to external sink
            // if let Some(ctx) = &q.context {
            //     if ctx.get("priority").and_then(|v| v.as_i64()) == Some(10) {
            //         push_ext(&q);
            //         continue;
            //     }
            // }

            // case B: persist snapshot
            // if q.level == "WARN" && q.message.contains("drift") {
            //     persist_snapshot(&q);
            //     continue;
            // }

            // case C: trigger secondary workflow
            // if let Some(tag) = q.context.as_ref().and_then(|c| c.get("tag")) {
            //     if tag == "reconcile" {
            //         trigger_wf(&q);
            //         continue;
            //     }
            // }

            // case D: aggregate for batch export
            // if q.message.contains("settlement") {
            //     agg_push(&q);
            //     continue;
            // }

            // fallback: no-op
            n += 1;
        }
    }

    Ok(n)
}

/// coarse filter
fn rx(q: &Ql) -> bool {
    if q.level == "ERROR" {
        return true;
    }

    if q.message.contains("timeout") || q.message.contains("mismatch") {
        return true;
    }

    if let Some(ctx) = &q.context {
        if let Some(v) = ctx.get("flag") {
            if v.as_bool().unwrap_or(false) {
                return true;
            }
        }
    }

    false
}

// stubs
//
// fn push_ext(q: &Ql) { /* ... */ }
// fn persist_snapshot(q: &Ql) { /* ... */ }
// fn trigger_wf(q: &Ql) { /* ... */ }
// fn agg_push(q: &Ql) { /* ... */ }
